module/shop_beta/base.py

```diff
@@ -0,0 +1,324 @@
+            items = shop_items.items
+            known = len([item for item in items if item.is_known_item])
+            logger.attr('Item detected', known)
+            if known == 0 or known != record:
+                record = known
+                continue
+            else:
+                record = known
+
+            # End
+            if self.shop_has_loaded(items):
+                break
+
+        # Log final result on predicted items
+        items = shop_items.items
+        grids = shop_items.grids
+        if len(items):
+            min_row = grids[0, 0].area[1]
+            row = [str(item) for item in items if item.button[1] == min_row]
+            logger.info(f'Shop row 1: {row}')
+            row = [str(item) for item in items if item.button[1] != min_row]
+            logger.info(f'Shop row 2: {row}')
+            return items
+        else:
+            logger.info('No shop items found')
+            return []
+
+    def shop_check_item(self, item):
+        """
+        Override in variant class
+        for specific check item
+        actions
+
+        Args:
+            item: Item to check
+
+        Returns:
+            bool:
+        """
+        return False
+
+    def shop_check_custom_item(self, item):
+        """
+        Override in variant class
+        for specific check custom item
+        actions; no restriction to filter string
+
+        Args:
+            item (Item):
+
+        Returns:
+            bool:
+        """
+        return False
+
+    def shop_interval_clear(self):
+        """
+        Override in variant class
+        if need to clear particular
+        asset intervals
+        """
+        pass
+
+    def shop_buy_handle(self, item):
+        """
+        Override in variant class
+        for specific buy handle
+        actions
+
+        Args:
+            item (Item):
+
+        Returns:
+            bool:
+        """
+        return False
+
+    def shop_get_item_to_buy(self, items, selection=''):
+        """
+        Args:
+            items list(Item): acquired from shop_get_items
+            selection (str): user configured value, items desired
+
+        Returns:
+            Item: Item to buy, or None.
+        """
+        # Convert selection to list{str], remove any empty/whitespace entries
+        selection = selection.replace(' ', '').replace('\n', '').split('>')
+        selection = list(filter(''.__ne__, selection))
+
+        for select in selection:
+            # 'Choice Ship' purchases are not supported
+            if 'ship' in select.lower():
+                continue
+
+            for item in items:
+                if self.shop_check_custom_item(item):
+                    return item
+                if select not in item.alt_name:
+                    continue
+                if not self.shop_check_item(item):
+                    continue
+
+                return item
+
+        return None
```

Comment:
Will need help on this one, having difficulty visualizing the changes needed for this and running into walls I'm not sure how to approach.  I've jotted some notes down while trying to review the Filter class and how research module uses it. https://pastebin.com/zVkQLymd

Reply:
Had some time so did a deep read on `Filter` class more closely and refreshing how `re` module functions. Initially had an incorrect approach with the regex so I think I've got it down with this.
https://pastebin.com/DERbLfLD

But I'm not sure whether I would consider this an improvement from the current code as this regex can allow for bad filter strings to pass the condition.  Unless I again don't have the right regex or am over thinking this.

Reply:
> Can omit the `select not in item.alt_name` check as done in original since already handled by Filter using regexp?

Yes

> Are there other attributes that can be used other than 'name'?

Yes, and it's recommended to. All problems should be solved by adding attributes. Take the research filter as an example, image  detection only gives 2 attributes, `series` and `name`, but they are parsed into 5, `series`, `ship`, `ship_rarity`, `genre`, `duration`. So the item names of shops should be parsed in to 4 attributes after the image detection, `group`, `item`, `sub_genre`, `tier`.

Well, I've just wrote here when you gave a newer reply, I'll keep the above although they seem to be outdated.

`prredt2` will pass but won't select anything, because there‘s no such an item can match it. Most part of Alas is assuming user will give a proper input, cause protecting Alas from someone intended to mess it up will add myself a lot of work.

Reply:
Gotcha. After some more mulling, double checking the classes, and your comments. I think the earlier issues I was stuck on I've gone over the hill but I think I ran into others as a kept going. 

Probably am just overthinking these things or blind so sorry for that but these are my latest compiled notes.
https://pastebin.com/hwZUjJWc

Doesn't feel like `predict()` got any less complicated though, still need to be careful setting up the attribute values

Also I wasn't sure how to break the item names into 4 attribute values since the naming convention I had made seems at most 3 can be broken down into. Could you provide examples?
`BookRedT3 --> Book, Red, T3; RetrofitT2 --> Retrofit, None, T2; PRNeptuneBP --> PR, Neptune, BP`

Reply:
You could use the same regular expression to match the item names from predict().
4 attributes are, group, item, sub_genre, tier. Both 3 and 4 are ok, choose the one you like.
> group, an upper classification of items: DR, PRY
> item, name of items: Book, Retrofit, Neptune
> sub_genre, a lower classification of items: Red, Blue
> tier, another categorical dimension: T1, T2